File: claude-plugin/scripts/format_status.php
#!/usr/bin/env php
<?php
/**
 * Parse execution JSON and output a readable status summary.
 *
 * Usage: buddy executions:show <id> --pipeline=<id> --json | php format_status.php
 *
 * Highlights failed actions with clear indicators.
 */

function parseTime(?string $ts): ?DateTimeImmutable
{
    if (!$ts) {
        return null;
    }
    try {
        return new DateTimeImmutable($ts);
    } catch (Exception $e) {
        return null;
    }
}

function formatDuration(?string $start, ?string $end): string
{
    $startDt = parseTime($start);
    $endDt = parseTime($end);
    if (!$startDt || !$endDt) {
        return '-';
    }
    $diff = $endDt->getTimestamp() - $startDt->getTimestamp();
    if ($diff < 60) {
        return "{$diff}s";
    }
    $minutes = intdiv($diff, 60);
    $seconds = $diff % 60;
    return "{$minutes}m {$seconds}s";
}

function statusIndicator(string $status): string
{
    $indicators = [
        'SUCCESSFUL' => '[OK]',
        'FAILED' => '[X]',
        'INPROGRESS' => '[->]',
        'ENQUEUED' => '[..]',
        'SKIPPED' => '[--]',
        'TERMINATED' => '[##]',
    ];
    return $indicators[$status] ?? '[?]';
}

function main(): int
{
    $input = file_get_contents('php://stdin');
    $data = json_decode($input, true);

    if (json_last_error() !== JSON_ERROR_NONE) {
        fwrite(STDERR, "ERROR: Invalid JSON input: " . json_last_error_msg() . "\n");
        return 1;
    }

    // Handle both single execution and array
    if (isset($data[0])) {
        if (count($data) === 0) {
            echo "No executions found.\n";
            return 0;
        }
        $data = $data[0];
    }

    // Execution header
    $execId = $data['id'] ?? 'Unknown';
    $status = $data['status'] ?? 'UNKNOWN';
    $branch = $data['branch']['name'] ?? '-';
    $revision = substr($data['to_revision']['revision'] ?? '-', 0, 8);
    $creator = $data['creator']['name'] ?? '-';
    $duration = formatDuration($data['start_date'] ?? null, $data['finish_date'] ?? null);

    echo "Execution #{$execId}\n";
    echo "  Status:   " . statusIndicator($status) . " {$status}\n";
    echo "  Branch:   {$branch}\n";
    echo "  Revision: {$revision}\n";
    echo "  Creator:  {$creator}\n";
    echo "  Duration: {$duration}\n";
    echo "\n";

    // Actions summary
    $actions = $data['action_executions'] ?? [];
    if (empty($actions)) {
        echo "No actions in this execution.\n";
        return 0;
    }

    $failed = array_filter($actions, fn($a) => ($a['status'] ?? '') === 'FAILED');
    $successful = array_filter($actions, fn($a) => ($a['status'] ?? '') === 'SUCCESSFUL');

    $total = count($actions);
    $passedCount = count($successful);
    echo "Actions: {$passedCount}/{$total} passed\n\n";

    // List all actions with status
    foreach ($actions as $action) {
        $name = $action['action']['name'] ?? 'Unknown';
        $actionStatus = $action['status'] ?? 'UNKNOWN';
        $actionDuration = formatDuration($action['start_date'] ?? null, $action['finish_date'] ?? null);
        $indicator = statusIndicator($actionStatus);

        $line = "  {$indicator} {$name}";
        if ($actionDuration !== '-') {
            $line .= " ({$actionDuration})";
        }
        echo $line . "\n";
    }

    // Highlight failures
    if (!empty($failed)) {
        echo "\nFAILED ACTIONS:\n";
        foreach ($failed as $action) {
            $name = $action['action']['name'] ?? 'Unknown';
            $type = $action['action']['type'] ?? '-';
            echo "  - {$name} ({$type})\n";
        }
    }

    return 0;
}

exit(main());

File: scripts/format_status.py
#!/usr/bin/env python3
"""Parse execution JSON and output a readable status summary.

Usage: buddy executions:show <id> --pipeline=<id> --json | python format_status.py

Highlights failed actions with clear indicators.
"""

import json
import sys
from datetime import datetime


def parse_time(ts):
    """Parse ISO timestamp to datetime, return None if invalid."""
    if not ts:
        return None
    try:
        return datetime.fromisoformat(ts.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return None


def format_duration(start, end):
    """Calculate and format duration between two timestamps."""
    start_dt = parse_time(start)
    end_dt = parse_time(end)
    if not start_dt or not end_dt:
        return "-"
    delta = end_dt - start_dt
    total_seconds = int(delta.total_seconds())
    if total_seconds < 60:
        return f"{total_seconds}s"
    minutes, seconds = divmod(total_seconds, 60)
    return f"{minutes}m {seconds}s"


def status_indicator(status):
    """Return visual indicator for status."""
    indicators = {
        "SUCCESSFUL": "✓",
        "FAILED": "✗",
        "INPROGRESS": "→",
        "ENQUEUED": "○",
        "SKIPPED": "⊘",
        "TERMINATED": "□",
    }
    return indicators.get(status, "?")


def main():
    try:
        data = json.load(sys.stdin)
    except json.JSONDecodeError as e:
        print(f"ERROR: Invalid JSON input: {e}", file=sys.stderr)
        sys.exit(1)

    # Handle both single execution and array
    if isinstance(data, list):
        if len(data) == 0:
            print("No executions found.")
            sys.exit(0)
        data = data[0]

    # Execution header
    exec_id = data.get("id", "Unknown")
    status = data.get("status", "UNKNOWN")
    branch = data.get("branch", {}).get("name", "-")
    revision = data.get("to_revision", {}).get("revision", "-")[:8]
    creator = data.get("creator", {}).get("name", "-")
    duration = format_duration(data.get("start_date"), data.get("finish_date"))

    print(f"Execution #{exec_id}")
    print(f"  Status:   {status_indicator(status)} {status}")
    print(f"  Branch:   {branch}")
    print(f"  Revision: {revision}")
    print(f"  Creator:  {creator}")
    print(f"  Duration: {duration}")
    print()

    # Actions summary
    actions = data.get("action_executions", [])
    if not actions:
        print("No actions in this execution.")
        sys.exit(0)

    failed = [a for a in actions if a.get("status") == "FAILED"]
    successful = [a for a in actions if a.get("status") == "SUCCESSFUL"]

    print(f"Actions: {len(successful)}/{len(actions)} passed")
    print()

    # List all actions with status
    for action in actions:
        name = action.get("action", {}).get("name", "Unknown")
        action_status = action.get("status", "UNKNOWN")
        action_duration = format_duration(action.get("start_date"), action.get("finish_date"))
        indicator = status_indicator(action_status)

        line = f"  {indicator} {name}"
        if action_duration != "-":
            line += f" ({action_duration})"
        print(line)

    # Highlight failures
    if failed:
        print()
        print("FAILED ACTIONS:")
        for action in failed:
            name = action.get("action", {}).get("name", "Unknown")
            action_type = action.get("action", {}).get("type", "-")
            print(f"  - {name} ({action_type})")


if __name__ == "__main__":
    main()
